aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs

```diff
@@ -0,0 +1,216 @@
+        output_path: String,  // output path
+        verification_f: F,    // Modified to return a HashMap for multi-table data
+    ) -> anyhow::Result<HashMap<String, serde_json::Value>>
+    // Return HashMap for multi-table results
+    where
+        F: FnOnce(&str) -> anyhow::Result<HashMap<String, serde_json::Value>>
+            + Send
+            + Sync
+            + 'static, // Modified for multi-table verification
+    {
+        let retry_strategy = ExponentialBackoff::from_millis(100).map(jitter).take(5); // Retry up to 5 times
+
+        let timeout_duration = TokioDuration::from_secs(10); // e.g., 5 seconds timeout for each retry
+        let result = Retry::spawn(retry_strategy, || async {
+            // Wrap processor call with a timeout
+            match time::timeout(timeout_duration, processor.run_processor()).await {
+                Ok(result) => result.context("Processor run failed"),
+                Err(_) => Err(anyhow::anyhow!("Processor run timed out")),
+            }
+        })
+        .await;
+
+        // Handle failure after retries
+        match result {
+            Ok(_) => {
+                println!("[INFO] Processor run succeeded");
+            },
+            Err(e) => {
+                eprintln!("[ERROR] Processor failed after retries: {:?}", e);
+                return Err(anyhow::anyhow!(
+                    "Failed to run processor after multiple retries: {}",
+                    e
+                ));
+            },
+        }
+
+        // Small delay to ensure all data is processed before verification
+        tokio::time::sleep(Duration::from_millis(250)).await;
+
+        // Retrieve data from multiple tables using verification function
+        let db_values = verification_f(db_url).context("Verification function failed")?;
+
+        // Conditionally generate output files for each table
+        if generate_files {
+            println!("[TEST] Generating output files for all tables.");
+
+            // Iterate over each table's data in the HashMap and generate an output file
+            for (table_name, table_data) in &db_values {
+                generate_output_file(
+                    processor.name(),
+                    table_name,
+                    &format!("{}", txn_version),
+                    table_data,
+                    output_path.clone(),
+                )?;
+            }
+        } else {
+            println!("[TEST] Skipping file generation as requested.");
+        }
+
+        // Return the HashMap containing the data from all queried tables
+        Ok(db_values)
+    }
+
+    /// Helper function to set up and run the mock GRPC server.
+    async fn setup_mock_grpc(
+        &self,
+        transactions_response: Vec<TransactionsResponse>,
+        chain_id: u64,
+    ) {
+        let mock_grpc_server = MockGrpcServer {
+            transactions_response: Mutex::new(transactions_response),
```

Comment:
Why does this need a Mutex around the txn's?

Reply:
this is. no longer needed as dynamic porting feature is added to mock grpc